k2/csrc/host_shim.h

```diff
@@ -179,6 +182,163 @@ class FsaVecCreator {
   int32_t next_fsa_idx_;
 };
 
+/*
+  Wrap some property test function in `host/properties.h`, they are generally
+  for test purpose for now and work only for CPU. Users usually would not call
+  these functions. Instead, Call `GetFsaBasicProperties` or
+  `GetFsaVecBasicProperties` in fsa.h if you want to check Fsa/FsaVec's
+  properites in production code.
+*/
+
+/*
+  Check properties of FsaOrVec (which must be on CPU) with property test
+  function `f` which is one of property test functions for k2host::Fsa in
+  host/properties.h, e.g. IsValid(const k2::host Fsa&), IsTopSorted(const
+  k2host::Fsa&).
+
+  If `fsas` is FsaVec, the function will return an array on CPU which has
+    ans[i] = f(FsaVecToHostFsa(fsa_vec, i)) for 0 <= i < fsa_vec.Dim0();
+  else
+    returns a CPU array with size 1 and ans[0] = f(FsaToHostFsa(fsas))
+*/
+Array1<bool> CheckProperties(FsaOrVec &fsas, bool (*f)(const k2host::Fsa &));
+
+/*
+  ans[i]=true if `fsas[i]` is valid. An Fsa is valid if:
+  1. It is a valid Ragged array.
+  2. If it's not empty, it contains at least two states.
+     Noted empty fsa is valid.
+  3. Only arcs with symbol==-1 enter the final state.
+  4. There's no arcs leaving final_state.
+  5. All arcs leaving a state have a same src_state, which is the corresponding
+     idx0 in fsa, see index naming convention explained in utils.h.
```

Comment:
This can be idx01 if the input is a FsaVec.

Reply:
I explain it here starting with ` An Fsa is valid if`. Will add `idx1` (it's not idx01)